File: bluetooth_scanner.go
// bluetooth_scanner.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"regexp"
	"time"

	"github.com/go-ble/ble"
	"github.com/go-ble/ble/examples/lib/dev"
)

type DeviceInfo struct {
	Address   string   `json:"address"`
	Name      string   `json:"name"`
	RSSI      int      `json:"rssi"`
	Timestamp string   `json:"timestamp"`
	Services  []string `json:"services"`
}

type Scanner struct {
	timeout      time.Duration
	filterRegex  *regexp.Regexp
	historyFile  string
	devices      []DeviceInfo
	history      []DeviceInfo
}

func NewScanner(timeout int, filter, history string) *Scanner {
	var re *regexp.Regexp
	if filter != "" {
		re = regexp.MustCompile(filter)
	}
	s := &Scanner{
		timeout:     time.Duration(timeout) * time.Second,
		filterRegex: re,
		historyFile: history,
	}
	s.loadHistory()
	return s
}

func (s *Scanner) loadHistory() {
	if s.historyFile == "" {
		return
	}
	data, err := ioutil.ReadFile(s.historyFile)
	if err != nil {
		if os.IsNotExist(err) {
			return
		}
		log.Fatal(err)
	}
	json.Unmarshal(data, &s.history)
}

func (s *Scanner) saveHistory() {
	if s.historyFile == "" {
		return
	}
	data, _ := json.MarshalIndent(s.history, "", "  ")
	ioutil.WriteFile(s.historyFile, data, 0644)
}

func (s *Scanner) filter(device ble.Device) bool {
	if s.filterRegex == nil {
		return true
	}
	return s.filterRegex.MatchString(device.Name())
}

func (s *Scanner) scan() {
	fmt.Printf("Scanning for %v...\n", s.timeout)
	d, err := dev.NewDevice("default")
	if err != nil {
		log.Fatal(err)
	}
	ble.SetDefaultDevice(d)

	// Scan with handler
	ch := make(chan bool)
	ble.Scan(context.Background(), true, func(adv ble.Advertisement) {
		if s.filter(adv) {
			info := DeviceInfo{
				Address:   adv.Addr().String(),
				Name:      adv.LocalName(),
				RSSI:      adv.RSSI(),
				Timestamp: time.Now().Format(time.RFC3339),
				Services:  adv.Services(),
			}
			s.devices = append(s.devices, info)
		}
	}, nil)

	time.Sleep(s.timeout)
	ble.StopScan()
	fmt.Printf("Found %d devices:\n", len(s.devices))
	for _, d := range s.devices {
		fmt.Printf("  %s (%s) RSSI: %d dBm\n", d.Name, d.Address, d.RSSI)
	}
	if s.historyFile != "" {
		s.history = append(s.history, s.devices...)
		s.saveHistory()
		fmt.Printf("History updated in %s\n", s.historyFile)
	}
}

func main() {
	timeout := flag.Int("timeout", 10, "scan duration in seconds")
	filter := flag.String("filter", "", "regex filter for device name")
	history := flag.String("history", "", "JSON history file")
	export := flag.String("export", "", "CSV export (requires --history)")
	flag.Parse()

	scanner := NewScanner(*timeout, *filter, *history)
	scanner.scan()

	if *export != "" && *history != "" {
		// Write CSV
		csvFile, _ := os.Create(*export)
		defer csvFile.Close()
		csvFile.WriteString("timestamp,name,address,rssi,services\n")
		for _, e := range scanner.history {
			line := fmt.Sprintf("%s,%s,%s,%d,\"%s\"\n", e.Timestamp, e.Name, e.Address, e.RSSI, strings.Join(e.Services, ";"))
			csvFile.WriteString(line)
		}
		fmt.Printf("History exported to %s\n", *export)
	}
}

File: bluetooth_scanner.js
// bluetooth_scanner.js
const noble = require('@abandonware/noble');
const fs = require('fs');
const { program } = require('commander');
const { performance } = require('perf_hooks');

class BluetoothScanner {
    constructor(options) {
        this.timeout = options.timeout || 10;
        this.filterPattern = options.filter ? new RegExp(options.filter) : null;
        this.historyFile = options.history || null;
        this.devices = [];
        this.history = this.loadHistory();
        this.startTime = null;
    }

    loadHistory() {
        if (this.historyFile && fs.existsSync(this.historyFile)) {
            try {
                return JSON.parse(fs.readFileSync(this.historyFile));
            } catch (e) {
                return [];
            }
        }
        return [];
    }

    saveHistory() {
        if (this.historyFile) {
            fs.writeFileSync(this.historyFile, JSON.stringify(this.history, null, 2));
        }
    }

    deviceFilter(device) {
        if (this.filterPattern && device.localName) {
            return this.filterPattern.test(device.localName);
        }
        return true;
    }

    startScan() {
        return new Promise((resolve) => {
            console.log(`Scanning for ${this.timeout} seconds...`);
            this.startTime = performance.now();
            noble.on('discover', (device) => {
                if (this.deviceFilter(device)) {
                    const info = {
                        address: device.address,
                        name: device.localName || 'Unknown',
                        rssi: device.rssi,
                        timestamp: new Date().toISOString(),
                        services: device.advertisement.serviceUuids || []
                    };
                    this.devices.push(info);
                }
            });

            noble.startScanning([], true);
            setTimeout(() => {
                noble.stopScanning();
                console.log(`Found ${this.devices.length} devices:`);
                this.devices.forEach(d => {
                    console.log(`  ${d.name} (${d.address}) RSSI: ${d.rssi} dBm`);
                });
                // Save history
                if (this.historyFile) {
                    this.history = this.history.concat(this.devices);
                    this.saveHistory();
                    console.log(`History updated in ${this.historyFile}`);
                }
                resolve(this.devices);
            }, this.timeout * 1000);
        });
    }
}

// Command line arguments
program
    .option('-t, --timeout <seconds>', 'Scan duration', parseInt, 10)
    .option('-f, --filter <regex>', 'Filter by device name')
    .option('-h, --history <file>', 'History JSON file')
    .option('-e, --export <file>', 'Export history to CSV (requires --history)')
    .parse(process.argv);

const opts = program.opts();
const scanner = new BluetoothScanner(opts);
scanner.startScan().then(() => {
    if (opts.export && opts.history) {
        const csvLines = ['timestamp,name,address,rssi,services'];
        scanner.history.forEach(entry => {
            csvLines.push(`${entry.timestamp},${entry.name},${entry.address},${entry.rssi},"${entry.services.join(';')}"`);
        });
        fs.writeFileSync(opts.export, csvLines.join('\n'));
        console.log(`History exported to ${opts.export}`);
    }
});
